internal/providers/vault/approle.go

```diff
@@ -0,0 +1,75 @@
+// Copyright 2022 Namespace Labs Inc; All rights reserved.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+
+package vault
+
+import (
+	"context"
+
+	vaultclient "github.com/hashicorp/vault-client-go"
+	"github.com/hashicorp/vault-client-go/schema"
+	"golang.org/x/sync/errgroup"
+	"namespacelabs.dev/foundation/framework/secrets"
+	"namespacelabs.dev/foundation/internal/fnerrors"
+	"namespacelabs.dev/foundation/std/tasks"
+	"namespacelabs.dev/foundation/universe/vault"
+)
+
+func (p *provider) appRoleProvider(ctx context.Context, secretId secrets.SecretIdentifier, cfg *vault.AppRole) ([]byte, error) {
+	vp := cfg.GetProvider()
+	if vp == nil {
+		return nil, fnerrors.BadInputError("invalid vault app role configuration: missing provider configuration")
+	}
+
+	vaultClient, err := p.Login(ctx, vp, vaultJwtAudience)
+	if err != nil {
+		return nil, err
+	}
+
+	if secretId.ServerRef == nil {
+		return nil, fnerrors.BadDataError("required server reference is not set")
+	}
+
+	return p.CreateSecretId(ctx, vaultClient, cfg)
+}
+
+func (p *provider) CreateSecretId(ctx context.Context, vaultClient *vaultclient.Client, cfg *vault.AppRole) ([]byte, error) {
+	return tasks.Return(ctx, tasks.Action("vault.create-secret-id").Arg("name", cfg.GetName()),
+		func(ctx context.Context) ([]byte, error) {
+			creds := vault.Credentials{
+				VaultAddress:   cfg.Provider.GetAddress(),
+				VaultNamespace: cfg.Provider.GetNamespace(),
+			}
+			wmp := vaultclient.WithMountPath(cfg.GetAuthMethod())
+
+			g := errgroup.Group{}
```

Comment:
Please use errgroup.WithContext to respect cancelation.

Or better yet, use foundation's executor.New()

Reply:
![image](https://github.com/namespacelabs/foundation/assets/196617/f9821c55-0819-4ca9-b24e-3901ad9c3048)

Much better.